cmd/recorder.go

```diff
@@ -0,0 +1,308 @@
+	// Ignore raw audio samples if client initiation is still in progress
+	if r.JackBufferSize <= 0 || r.JackSampleRate <= 0 {
+		return 0
+	}
+	// Read audio data from ports and immediately send to the channel for further processing
+	raw := [][]jack.AudioSample{}
+	for _, port := range r.RecorderPorts {
+		samples := port.GetBuffer(nframes)
+		raw = append(raw, samples)
+	}
+	r.RawSamplesChan <- raw
+	return 0
+}
+
+// onShutdown only runs upon unexpected connection error
+func (r *Recorder) onShutdown() {
+	r.ClientLock.Lock()
+	defer r.ClientLock.Unlock()
+	r.JackClient, r.RecorderPorts, r.JackSampleRate, r.JackBufferSize = nil, nil, 0, 0
+	for _, trash := range AudioFilenames {
+		cleanStaleFile(trash)
+	}
+	cleanStaleFile(filepath.Join(MediaDir, HLSIndex))
+	cleanStaleFile(filepath.Join(MediaDir, "init.mp4"))
+	AudioFilenames, FrameBuffer = nil, nil
+	HLSPlaylistHash = ""
+}
+
+// TeardownClient closes the currently active JACK client
+func (r *Recorder) TeardownClient() {
+	r.ClientLock.Lock()
+	defer r.ClientLock.Unlock()
+	if r.JackClient != nil {
+		r.JackClient.Close()
+	}
+	r.JackClient, r.RecorderPorts, r.JackSampleRate, r.JackBufferSize = nil, nil, 0, 0
+	for _, trash := range AudioFilenames {
+		cleanStaleFile(trash)
+	}
+	cleanStaleFile(filepath.Join(MediaDir, HLSIndex))
+	cleanStaleFile(filepath.Join(MediaDir, "init.mp4"))
+	cleanStaleFile(filepath.Join(MediaDir, "playlist-*.m3u8"))
+	AudioFilenames, FrameBuffer = nil, nil
+	HLSPlaylistHash = ""
+	log.Info("Teardown of JACK client completed")
+}
+
+// SetupClient establishes a new client to listen in on JACK ports
+func (r *Recorder) SetupClient() {
+	var err error
+	r.ClientLock.Lock()
+	defer r.ClientLock.Unlock()
+	err = waitForDaemon()
+	if err != nil {
+		panic(err)
+	}
+	portRegistrationFunc := func(client *jack.Client) {
+		for i := 1; i <= NumRecorderChannels; i++ {
+			portName := fmt.Sprintf("send_%d", i)
+			portIn := client.PortRegister(portName, jack.DEFAULT_AUDIO_TYPE, jack.PortIsInput, 0)
+			r.RecorderPorts = append(r.RecorderPorts, portIn)
+		}
+	}
+	client, err := initClient(r.Name, nil, r.onShutdown, r.processBuffer, portRegistrationFunc, false)
+	if err != nil {
+		panic(err)
+	}
+	r.JackClient = client
+	r.JackSampleRate = int(r.JackClient.GetSampleRate())
+	r.JackBufferSize = int(r.JackClient.GetBufferSize())
+	HLSPlaylistHash = GenerateRandomString(8)
+	log.Info("Setup of JACK client completed", "name", r.JackClient.GetName())
+}
+
+// Run is the primary loop that is connects new JACK ports upon registration
+func (r *Recorder) Run(wg *sync.WaitGroup) {
+	defer wg.Done()
+	for {
+		raw, ok := <-r.RawSamplesChan
+		if !ok {
+			log.Info("Raw samples channel is closed")
+			return
+		}
+		r.addFrame(raw)
+	}
+}
+
+func (r *Recorder) addFrame(audioSamples [][]jack.AudioSample) {
+	// Get current sample rate + buffer size in a thread-safe manner in case the JACK client fails
+	r.ClientLock.Lock()
+	sampleRate, bufferSize := r.JackSampleRate, r.JackBufferSize
```

Comment:
Would these ever change after initial setup of the client? Locking is expensive from a perf perspective, so we should try to eliminate this if we can.

Reply:
Yeah this can be removed